### ISSUE TITLE
TWEAK doesn't work in inherited classes

### ISSUE THREAD

**Issue by wayland** [lizmat/Hash-Agnostic#3]
This class overrides new and doesn't do anything with TWEAK/BUILD.  This means that TWEAK and BUILD don't work in inherited classes.  Is this something that you're willing to fix?  

Maybe "callsame" would be useful?  

Thanks!

**Comment by wayland** [comment]
It also seems to be breaking variable initialisation.

**Comment by lizmat** [comment]
Could you supply me with an example that breaks for you?

**Comment by wayland** [comment]
I'd like to, but the following gives me a segfault:

```
#!/usr/bin/raku

use v6.c;

use     Hash::Agnostic:ver<0.0.11>:auth<zef:lizmat>;

class   FooTest does Hash::Agnostic {
        submethod TWEAK() {
                say "hi";
        }
}

my $foo = FooTest.new();

say "done";
```

That's probably because my Raku isn't up-to-date.  

So then I tried:
```
#!/usr/bin/raku

use v6.c;

use     Hash::Agnostic:ver<0.0.11>:auth<zef:lizmat>;

class   FooTest does Hash::Agnostic {
        submethod TWEAK() {
                say "TWEAK";
        }

        method  AT-KEY($key) {
                say "AT-KEY";
                return 1;
        }
        method  keys() {
                say "keys";
                return 0;
        }
}

my $foo = FooTest.new();

say "done";
```
...and I got 
```
keys
No implementation of DELETE-KEY method found for FooTest.
  in method CLEAR at /home/wayland/.raku/sources/0740704BCA695A7F109F9715ADF8887B91523B85 (Hash::Agnostic) line 54
  in method STORE at /home/wayland/.raku/sources/0740704BCA695A7F109F9715ADF8887B91523B85 (Hash::Agnostic) line 62
  in method new at /home/wayland/.raku/sources/0740704BCA695A7F109F9715ADF8887B91523B85 (Hash::Agnostic) line 99
  in block <unit> at ./tweak-broken-demo line 22
```
...and I wasn't expecting to implement DELETE-KEY.  

So I tried 
```
#!/usr/bin/raku

use v6.c;

use     Hash::Agnostic:ver<0.0.11>:auth<zef:lizmat>;

class   FooTest does Hash::Agnostic {
        has     Str     $.color = 'red';

        submethod TWEAK() {
                say "TWEAK";
        }

        method  AT-KEY($key) {
                say "AT-KEY";
                return 1;
        }
        method  keys() {
                say "keys";
                return 0;
        }

        method  DELETE-KEY($key) {
                say "deleted";
        }
}

my $foo = FooTest.new();

say $foo.color;

say "done";
```
...and I got 
```
keys
deleted
(Str)
done
```
Note that at no point does the above output say "TWEAK", and when it should say "red", it says "(Str)".  

Thanks!

**Comment by wayland** [comment]
Is the final example above a clear enough example to explain the problem?

**Comment by lizmat** [comment]
Thanks for the nudge.  I've released 0.13 now.

It appears that having a `.list` or a `.List` method in the role triggers something deep in the bowels of role composition, sadly without backtrace.

I don't have the mindset to go into the bowels atm, so I decided to remove default support for the `.list`and `.List` methods for now.

Please confirm this fixes the issue for you!  :-)   Or not  :-(

**Comment by wayland** [comment]
Unfortunately this doesn't seem to have altered things at all -- the final demo above still produces the same erroneous output.

**Comment by lizmat** [comment]
Did you change the `use` statement?

**Comment by wayland** [comment]
I did indeed.  

I've just put through a PR for a couple of tests that should work, but don't.  Feel free to use them however you see fit.

HTH,

**Comment by lizmat** [comment]
Thanks for the PR, that was helpful.

Version 0.0.14 now in the ecosystem.  The final fix was pretty simple: `s/CREATE/bless(|%_)/` basically  :-)

**Comment by lizmat** [comment]
@wayland I was intrigued by your use of indentation here:
```
use     Hash::Agnostic:ver<0.0.11>:auth<zef:lizmat>;

class   FooTest does Hash::Agnostic {
        has     Str     $.color = 'red';
```
This feels very PLATO / TenCORE like.  OOC, did you by any chance work on those systems in the past?

**Comment by wayland** [comment]
@lizmat 

Regarding the fix, great!  I had a feeling it would be something straightforward, but wasn't expecting it to be quite that easy :) . I can confirm that it works for me.  

Regarding PLATO / TenCORE, no, I haven't.  I think I'm influence by a strong natural bent towards tables, so I tend to lay things out in a tabular format where possible (though I love multitrees too).  I have also used tools in the past where the editor gave the user control over where the tab stops were, which probably increased that bent towards tables.  I use the tab character where possible; given the choice, I'd like to set tab stops so that I would only need one tab character to go to the next column, but that's not generally possible with modern editors.  

I started out on GWBASIC in primary school, with a smattering of x86 assembly and dBase in high school, then Pascal/Delphi (where I learned linked lists in year 11 by reading some code I found), and finally onto Perl while at Uni.  

One thing I *will* say -- I do tend to end up sooner or later using every feature available in a word processor, so not sure whether that influenced me or not (they *do* have adjustable tab stops).  

HTH,
